Tweak Quick Open theming

MewPurPur opened the pull request:
Addresses some, but not all points in: https://github.com/godotengine/godot-proposals/discussions/11558

The things I've fixed are the issues with spacing and sizing.

Changed up the structure of list elements to prevent thumbnails from stretching vertically. Here's how this looked before:

![image](https://github.com/user-attachments/assets/7b92ac89-8281-45af-afb6-8774c39e9899)

Changed up the structure of grid elements too, to give some space above and to have better spacing. Here's how it looked before:

![image](https://github.com/user-attachments/assets/360a26ec-ae38-4d3e-8005-aac5834d770d)

And there's now no spacing between elements, prevent you from accidentally clicking on the tiny space between.

I've also fixed the ugly hover style, which came from overriding the nonexistent "hover" style instead of "hovered".

Here's how it looks now:

![image](https://github.com/user-attachments/assets/19206a85-0ca9-4d1d-b89b-00e52091abcc)

![image](https://github.com/user-attachments/assets/9af4eab9-5057-4c2c-86d5-ba6f58f85a81)

Please tell me if you think the spacing is too tight now.

MewPurPur committed:
Tweak Quick Open theming

Calinou reviewed (commented):
Tested locally, it works as expected.

![image](https://github.com/user-attachments/assets/9993a634-0413-4cc5-9cba-06c1daeee785)

The amount of spacing looks good to me. However, the icons aren't centered in the hovered/selected area:

![image](https://github.com/user-attachments/assets/6dafd866-bc10-4b83-a02e-30821b44fcd3)

There are 6 pixels above the icon and 7 pixels below the icon. (The icon is 36 pixels tall.)
I can notice this issue across every item, so it's not specific to some items on the list.

I suggest adding 1 pixel of padding at the top so it's even.

MewPurPur commented:
Sure. This is because the vboxcontainer with the text is 37px tall, because text has this tendence of taking more space than what it seems like it should.

I'll fix it.

arkology commented on editor/gui/editor_quick_open_dialog.cpp:
```diff
@@ -932,43 +933,42 @@ static Vector2i _get_name_interval(const Vector2i &p_interval, int p_dir_index)
 QuickOpenResultListItem::QuickOpenResultListItem() {
 	set_h_size_flags(Control::SIZE_EXPAND_FILL);
-	add_theme_constant_override("separation", 4 * EDSCALE);
+	begin_bulk_theme_override();
+	add_theme_constant_override("margin_left", 6 * EDSCALE);
+	add_theme_constant_override("margin_right", 6 * EDSCALE);
+	end_bulk_theme_override();
```
If I remember correctly, using bulk functions is necessary only for runtime changed inside tree. So there is no need to use it in constructor. Same for `QuickOpenResultGridItem` constructor below.

MewPurPur commented:
Doned, bumping up the "anti-separation" of the two labels by 1 more pixel looks good to me even with the maximum ascents/descents in the default font.

Calinou reviewed (approved):
Looks great and even now :slightly_smiling_face: 

![image](https://github.com/user-attachments/assets/cac97f0d-b8f9-4e5f-b45e-c6ae5ca96c2e)

It's still a tad uneven at 200% editor scale (10 pixels above the icon, 11 pixels below) but it's probably unnoticeable on an actual hiDPI display.

![image](https://github.com/user-attachments/assets/76abd808-0edc-4bc7-a702-8c63deddc816)

MewPurPur commented:
Since the dialog redesign is new to 4.4, I think it'd be great to merge the theming fixes timely.

MewPurPur commented:
Okay, I added SNAME where you suggested, but I honestly still don't get why it was suggested only for some of the strings, while others don't have it (like "separation" outside of the lines I changed, "selected" next to my lines, or even "margin_*" constants inside the diff)

Ivorforce commented:
Here's what `SNAME` does: It statically assigns a `StringName` that will be referenced at this place for future use.
This is much faster than using String literals, but also forces the `StringName` data to be kept in memory. It is appropriate in places where the `StringName` data is expected to be kept in memory anyway, and especially when it is used in performance critical functions, like those called every tick.

I don't think there is an "official" recommendation for when to use it, but if i were to make one, I'd probably suggest using it for most static strings since most are expected to be kept in memory for one reason or another anyway. The cost of a reference to it is rather small.

AThousandShips commented:
The general recommendation is to use it in hot paths, and not to use it everywhere. In modern cases speed usually wins over memory when prioritising but there's no use in doing it in paths that are called exactly once in a run

arkology commented:
@Ivorforce thanks for explanation. So actually "if it was used multiple times in other places" is not necessary, it is enough if it is already used in any "common" path (i.e. already cached).

AThousandShips commented:
I think the important takeaway is that optimization (or any code pattern) should be applied with intentionality, with an understanding of why it's used, and that any pattern used "just because" is a code smell.

`SNAME` is a pretty harmless one but it's always good to get used to different patterns and why we use them and when

akien-mga commented:
Thanks!
